cephfs/directory.go

```diff
@@ -0,0 +1,131 @@
+package cephfs
+
+/*
+#cgo LDFLAGS: -lcephfs
+#cgo CPPFLAGS: -D_FILE_OFFSET_BITS=64
+#include <stdlib.h>
+#include <dirent.h>
+#include <cephfs/libcephfs.h>
+*/
+import "C"
+
+import (
+	"unsafe"
+)
+
+// Directory represents an open directory handle.
+type Directory struct {
+	mount *MountInfo
+	dir   *C.struct_ceph_dir_result
+}
+
+// OpenDir returns a new Directory handle open for I/O.
+//
+// Implements:
+//  int ceph_opendir(struct ceph_mount_info *cmount, const char *name, struct ceph_dir_result **dirpp);
+func (mount *MountInfo) OpenDir(path string) (*Directory, error) {
+	var dir *C.struct_ceph_dir_result
+
+	cPath := C.CString(path)
+	defer C.free(unsafe.Pointer(cPath))
+
+	ret := C.ceph_opendir(mount.mount, cPath, &dir)
+	if ret != 0 {
+		return nil, getError(ret)
+	}
+
+	return &Directory{
+		mount: mount,
+		dir:   dir,
+	}, nil
+}
+
+// Close the open directory handle.
+//
+// Implements:
+//  int ceph_closedir(struct ceph_mount_info *cmount, struct ceph_dir_result *dirp);
+func (dir *Directory) Close() error {
+	return getError(C.ceph_closedir(dir.mount.mount, dir.dir))
+}
+
+// Inode represents an innode number in the file system.
+type Inode uint64
+
+// DirEntry represents an entry within a directory.
+type DirEntry struct {
+	Inode Inode
```

Comment:
This probably should be a pointer to an Inode, an Inode can be part of multiple DirEntries (think hardlinks). Maybe there is no need to optimize for that, but it follows the traditional filesystem structure.

Assuming DirEntry will get extended with more (optional) attributes, it might be a good idea to prepare for that. Many filesystems use a `valid` mask that contains a bit for each attribute that was fetched. When an extended request for more attributes is done, more bits in the `valid` mask are set. In this case, I can imaging something like
```go
type DirEntry struct {
    valid uint64
    Inode Inode
    Name string
}

const (
    DirEntryInode = 0x1
    DirEntryName = 0x2
)

func toDirEntry(..) {
	return &DirEntry{
		valid: (DirEntryInode|DirEntryName),
		Inode: Inode(de.d_ino),
		Name:  C.GoString(&de.d_name[0]),
	}
}
```
And maybe use accessor functions like `GetInode()` and `GetName()` or provide an other way of checking that attributes are available.

Reply:
It's just an integer id, which will be copied (like a pointer value would). I'm ok with accessor functions rather than "direct" attributes as it allows the use of simple interfaces. However the rest of this seems unneeded as the fields are always going to be populated from the underlying direntry from C/ceph.
I will make the accessor change since I was thinking about doing that anyway.